src/PowerShellEditorServices/Services/PowerShell/Host/PsesInternalHost.cs

```diff
@@ -862,23 +875,53 @@ private void OnCancelKeyPress(object sender, ConsoleCancelEventArgs args)
             }
         }
 
+        private static readonly ConsoleKeyInfo s_nullKeyInfo = new(
+            keyChar: ' ',
+            ConsoleKey.DownArrow,
+            shift: false,
+            alt: false,
+            control: false);
+
         private ConsoleKeyInfo ReadKey(bool intercept)
         {
             // PSRL doesn't tell us when CtrlC was sent.
             // So instead we keep track of the last key here.
             // This isn't functionally required,
             // but helps us determine when the prompt needs a newline added
 
-            _lastKey = ConsoleProxy.SafeReadKey(intercept, CancellationToken.None);
-            return _lastKey.Value;
+            // NOTE: This requests that the client (the Code extension) send a non-printing key back
+            // to the terminal on stdin, emulating a user pressing a button. This allows
+            // PSReadLine's thread waiting on Console.ReadKey to return. Normally we'd just cancel
+            // this call, but the .NET API ReadKey is not cancellable, and is stuck until we send
+            // input. This leads to a myriad of problems, but we circumvent them by pretending to
+            // press a key, thus allowing ReadKey to return, and us to ignore it.
+            using CancellationTokenRegistration registration = _readKeyCancellationToken.Register(
+                () => _languageServer?.SendNotification("powerShell/sendKeyPress"));
```

Comment:
@SeeminglyScience I chatted with @andschwa about this line of code, and wonder will this cause multiple delegates that do the same thing gets registered to the token, which may end up sending multiple keys to terminal when the token is cancelled?
Would it be better to do this before passing the token to `ReadLine` method?

Reply:
The disposal (from `using`) after `ReadKey` *should* take care of that for us.

I worry that if we register it before `ReadKey` is actually called then we're setting yourselves up for a race condition. If it's registered at `ReadLine` and the cancellation happens before `ReadKey` is actually called, we'll end up successfully cancelling *and* also sending input that we will no longer be able to capture and throw away.

Reply:
That was my understanding as well.

Reply:
Thanks for the clarification. That makes sense!